### PR TITLE
Implement GetDeviceString on MacOS

### PR DESCRIPTION
**Pull request by X9VoiD**

### CODE REVIEW

File: HidSharp/Platform/MacOS/NativeMethods.cs [hunk 1]
```diff
@@ -383,6 +427,20 @@ public static string CFStringGetCharacters(IntPtr str)
...
+                                                                      UInt8 byte6, UInt8 byte7, UInt8 byte8, UInt8 byte9,
+                                                                      UInt8 byte10, UInt8 byte11, UInt8 byte12,
+                                                                      UInt8 byte13, UInt8 byte14, UInt8 byte15);
+
+        [DllImport(Foundation, EntryPoint = "CFUUIDGetUUIDBytes")]
+        public extern static CFUUIDBytes CFUUIDGetUUIDBytes(CFUUIDRef uuid);
+
+        [DllImport(IOKit, EntryPoint = "IOCreatePluginInterfaceForService")]
+        public extern static unsafe kern_return_t IOCreatePlugInInterfaceForService(io_service_t service, CFUUIDRef pluginType,
+                                                                                    CFUUIDRef interfaceType, ref IntPtr theInterface, ref SInt32 theScore);
```
Comment (InfinityGhost) [inline HidSharp/Platform/MacOS/NativeMethods.cs]:
```
Exception has occurred: CLR/System.EntryPointNotFoundException
Exception thrown: 'System.EntryPointNotFoundException' in HidSharpCore.dll: 'Unable to find an entry point named 'IOCreatePluginInterfaceForService' in shared library '/System/Library/Frameworks/IOKit.framework/IOKit'.'
   at HidSharp.Platform.MacOS.NativeMethods.IOCreatePlugInInterfaceForService(UInt32 service, IntPtr pluginType, IntPtr interfaceType, IntPtr& theInterface, Int32& theScore)
   at HidSharp.Platform.MacOS.MacHidDevice.GetDeviceString(Int32 index) in /Users/infinity/Documents/Repos/HidSharpCore/HidSharp/Platform/MacOS/MacHidDevice.cs:line 200
```

**Review by InfinityGhost — CHANGES_REQUESTED** [review]

File: HidSharp/Platform/MacOS/NativeMethods.cs (hunk 1 above)
Comment (InfinityGhost) [reply HidSharp/Platform/MacOS/NativeMethods.cs]: Typo, should be `PlugIn` not `Plugin`

File: HidSharp/Platform/MacOS/NativeMethods.cs [hunk 2]
```diff
@@ -495,6 +560,9 @@ public static byte[] IORegistryEntryGetCFProperty_Data(int entry, IntPtr dataKey
         [DllImport(IOKit, EntryPoint = "IOServiceMatching")] // name = IOHIDDevice
         public static extern IntPtr IOServiceMatching([MarshalAs(UnmanagedType.LPStr)] string name);
 
+        [DllImport(IOKit, EntryPoint = "QueryInterface")]
+        public static extern int QueryInterface(IntPtr a, IntPtr b, ref IntPtr c);
```
Comment (InfinityGhost) [inline HidSharp/Platform/MacOS/NativeMethods.cs]:
```
Exception has occurred: CLR/System.EntryPointNotFoundException
Exception thrown: 'System.EntryPointNotFoundException' in HidSharpCore.dll: 'Unable to find an entry point named 'QueryInterface' in shared library '/System/Library/Frameworks/IOKit.framework/IOKit'.'
   at HidSharp.Platform.MacOS.NativeMethods.QueryInterface(IntPtr a, IntPtr b, IntPtr& c)
   at HidSharp.Platform.MacOS.MacHidDevice.GetDeviceString(Int32 index) in /Users/infinity/Documents/Repos/HidSharpCore/HidSharp/Platform/MacOS/MacHidDevice.cs:line 207
```

Comment (InfinityGhost) [reply HidSharp/Platform/MacOS/NativeMethods.cs]: Maybe QueryInterface is reference to [Marshal.QueryInterface()](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.queryinterface?view=netcore-3.1)?

File: HidSharp/Platform/MacOS/NativeMethods.cs (hunk 1 above)
Comment (X9VoiD) [reply HidSharp/Platform/MacOS/NativeMethods.cs]: Typo fixed

File: HidSharp/Platform/MacOS/NativeMethods.cs (hunk 2 above)
Comment (X9VoiD) [reply HidSharp/Platform/MacOS/NativeMethods.cs]: New changes use Marshal now

File: HidSharp/Platform/MacOS/MacHidDevice.cs [hunk 3]
```diff
@@ -174,7 +177,69 @@ public override byte[] GetRawReportDescriptor()
...
+
+                    IOCreatePlugInInterfaceForService(deviceEntry,
+                        kIOUSBDeviceUserClientTypeID, kIOCFPlugInInterfaceID,
+                        ref iodev, ref score);
+                    if (iodev != null)
+                    {
+                        IntPtr devPtr = IntPtr.Zero;
+                        var deviceInterface = kIOUSBDeviceInterfaceGUID;
+
+                        Marshal.QueryInterface(iodev, ref deviceInterface, out devPtr);
```
Comment (InfinityGhost) [inline HidSharp/Platform/MacOS/MacHidDevice.cs:209]: Was hoping that the last reference was right, but apparently MacOS doesn't support COM Interop
```
Exception has occurred: CLR/System.PlatformNotSupportedException
Exception thrown: 'System.PlatformNotSupportedException' in System.Private.CoreLib.dll: 'COM Interop is not supported on this platform.'
   at System.Runtime.InteropServices.Marshal.QueryInterface(IntPtr pUnk, Guid& iid, IntPtr& ppv)
   at HidSharp.Platform.MacOS.MacHidDevice.GetDeviceString(Int32 index) in /Users/infinity/Documents/Repos/HidSharpCore/HidSharp/Platform/MacOS/MacHidDevice.cs:line 209
```

**Review by InfinityGhost — CHANGES_REQUESTED** [review]

**Comment by InfinityGhost** [comment]
We're going to redo this from scratch, so I'm closing this PR.

**Pull request closed without merging**